File: app/src/main/java/com/stho/software/nyota/universe/AbstractPlanet.java
package com.stho.software.nyota.universe;

import com.stho.software.nyota.utilities.Degree;
import com.stho.software.nyota.utilities.Formatter;
import com.stho.software.nyota.utilities.IconNameValueList;
import com.stho.software.nyota.utilities.Moment;

/**
 * Created by shoedtke on 31.08.2016.
 */
public abstract class AbstractPlanet extends AbstractSolarSystemElement {

    @Override
    public boolean isPlanet() {
        return true;
    }

    public void calculatePhase(Sun sun) {
        elongation = Degree.arcCos((sun.R * sun.R + R * R - r * r) / (2 * sun.R * R));
        FV = Degree.arcCos((r * r + R * R - sun.R * sun.R) / (2 * r * R));
        phase  =  (1 + Degree.cos(FV)) / 2;
    }

    public abstract void calculateMagnitude();

    @Override
    public IconNameValueList getDetails(Moment moment) {
        IconNameValueList details = super.getDetails(moment);

        if (position.prevSetTime != null)
            details.add(com.stho.software.nyota.R.drawable.sunset, "Set", position.prevSetTime, moment.getTimeZone());

        if (position.riseTime != null)
            details.add(com.stho.software.nyota.R.drawable.sunrise, "Rise", position.riseTime, moment.getTimeZone());

        if (position.setTime != null)
            details.add(com.stho.software.nyota.R.drawable.sunset, "Set ", position.setTime, moment.getTimeZone());

        if (position.nextRiseTime != null)
            details.add(com.stho.software.nyota.R.drawable.sunrise, "Rise", position.nextRiseTime, moment.getTimeZone());

        details.add(com.stho.software.nyota.R.drawable.star, "FV", Degree.fromDegree(FV));
        details.add(com.stho.software.nyota.R.drawable.star, "Phase", Formatter.df3.format(phase));
        details.add(com.stho.software.nyota.R.drawable.star, "Phase angle", Formatter.df0.format(phaseAngle));
        details.add(com.stho.software.nyota.R.drawable.star, "Magnitude", Formatter.df2.format(magn));
        details.add(com.stho.software.nyota.R.drawable.star, "Parallax", Formatter.df3.format(parallacticAngle));
        details.add(com.stho.software.nyota.R.drawable.star, "In south", position.inSouth, moment.getTimeZone());
        details.add(com.stho.software.nyota.R.drawable.star, "Culmination angle", Degree.fromDegree(position.culmination));

        return details;
    }

    protected double getHeightFor(Moment moment) {
        AbstractPlanet planet = getPlanetFor(moment);
        return planet.getHeight();
    }

    private double getHeight() {
        return position.altitude - H0();
    }

    protected double H0() {
         return -0.583;
    }

    protected abstract AbstractPlanet getPlanetFor(Moment moment);
}

File: app/src/main/java/com/stho/nyota/sky/universe/AbstractPlanet.kt
package com.stho.nyota.sky.universe

import com.stho.nyota.sky.utilities.*
import com.stho.nyota.sky.utilities.Degree.Companion.cos

/**
 * Created by shoedtke on 31.08.2016.
 */
abstract class AbstractPlanet(override val name: String) : AbstractSolarSystemElement() {

    override val key: String
        get() = toKey(name)

    override val isPlanet: Boolean
        get() = true

    override fun calculatePhase(sun: Sun) {
        elongation = Degree.arcCos((sun.R * sun.R + R * R - mr * mr) / (2 * sun.R * R))
        FV = Degree.arcCos((mr * mr + R * R - sun.R * sun.R) / (2 * mr * R))
        phase = (1 + cos(FV)) / 2
    }

    private val magnAsString: String
        get() = Formatter.df2.format(magn)

    abstract fun calculateMagnitude()

    override fun getBasics(moment: Moment): PropertyList =
        super.getBasics(moment).apply {
            if (position?.isUp == true) {
                add(com.stho.nyota.R.drawable.sunset, "Previous Set", position?.prevSetTime, moment.timeZone)
                add(com.stho.nyota.R.drawable.sunrise, "Rise", position?.riseTime, moment.timeZone)
                add(com.stho.nyota.R.drawable.sunset, "Set", position?.setTime, moment.timeZone)
                add(com.stho.nyota.R.drawable.sunrise, "Next Rise", position?.nextRiseTime, moment.timeZone)
            } else {
                add(com.stho.nyota.R.drawable.sunrise, "Previous Rise", position?.prevRiseTime, moment.timeZone)
                add(com.stho.nyota.R.drawable.sunset, "Set", position?.setTime, moment.timeZone)
                add(com.stho.nyota.R.drawable.sunrise, "Rise", position?.riseTime, moment.timeZone)
                add(com.stho.nyota.R.drawable.sunset, "Next Set", position?.nextSetTime, moment.timeZone)
            }
            add(com.stho.nyota.R.drawable.empty, "Magnitude", magnAsString)
            add(com.stho.nyota.R.drawable.empty, "In south", position?.inSouth, moment.timeZone)
            add(com.stho.nyota.R.drawable.empty, "Culmination angle", Degree.fromDegree(position!!.culmination))
        }

    override fun getDetails(moment: Moment): PropertyList =
        super.getDetails(moment).apply {
            add(com.stho.nyota.R.drawable.empty, "FV", Degree.fromDegree(FV))
            add(com.stho.nyota.R.drawable.empty, "Phase", Formatter.df3.format(phase))
            add(com.stho.nyota.R.drawable.empty, "Phase angle", Formatter.df0.format(phaseAngle))
            add(com.stho.nyota.R.drawable.empty, "Parallax", Formatter.df3.format(parallacticAngle))
        }

    override fun getHeightFor(moment: IMoment): Double =
        getPlanetFor(moment).height

    private val height: Double
        get() = position!!.altitude - H0()

    override fun H0(): Double =
        -0.583

    protected abstract fun getPlanetFor(moment: IMoment): AbstractPlanet

    companion object {

        private fun toKey(name: String): String =
            "PLANET:$name"

        fun isValidKey(key: String): Boolean =
            key.startsWith("PLANET:")

        fun nameFromKey(key: String): String =
            key.substring(7)
    }
}